{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e426ec21",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    "<ol>\n",
    " <li><a href=\"#introduction\">Introduction</a></li>\n",
    " <ol>\n",
    "  <li><a href=\"#using-the-notebook\">Using the notebook</a></li>\n",
    " </ol>\n",
    "    <li><a href=\"#the-proteins\">The proteins</a></li>   \n",
    "  <ol>\n",
    "    <li><a href=\"#1c8q\">1C8Q: Human salivary amylase</a></li>\n",
    "    <li><a href=\"#4m6u\">4M6U: *P. putida* mandelate racemase</a></li>\n",
    "    <li><a href=\"#4ads\">4ADS: Plasmodial PLP synthase complex</a></li>\n",
    "    <li><a href=\"#1ubq\">1UBQ: Ubiquitin (1.8 Å resolution)</a></li>\n",
    "    <li><a href=\"#1zqa\">1ZQA: DNA polymerase complexed with DNA</a></li>\n",
    "  </ol>\n",
    "    <li><a href=\"#the-code-cells\">The code cells</a></li>\n",
    "</ol>\n",
    "\n",
    "# Introduction <a class=\"anchor\" id=\"introduction\"></a>\n",
    "The Protein Data Bank (PDB) stores data for various proteins, each protein being designated by a four character alphanumeric ID. Some of this data includes 3-D structure information; this is important because the structure of a protein determines its function.\n",
    "\n",
    "So what is the PDB? Taken from their [website](https://pdb101.rcsb.org/learn/guide-to-understanding-pdb-data/introduction):\n",
    "\n",
    "> The PDB archive is a repository of atomic coordinates and other information describing proteins and other important biological macromolecules. Structural biologists use methods such as X-ray crystallography, NMR spectroscopy, and cryo-electron microscopy to determine the location of each atom relative to each other in the molecule. They then deposit this information, which is then annotated and publicly released into the archive by the wwPDB.\n",
    "\n",
    "Even still, predicting the shape of a protein remains challenging, though [huge breakthroughs](https://www.nature.com/articles/d41586-021-02025-4?error=cookies_not_supported&code=c3799b46-4355-4fca-9513-8d35f4ec4984) have been made in this field using [machine learning](https://pubmed.ncbi.nlm.nih.gov/34015749/). This notebook serves as a brief exploration into the world of proteins and protein modeling.\n",
    "\n",
    "### Using the notebook <a class=\"anchor\" id=\"using-the-notebook\"></a>\n",
    "Execute both of the [code cells](#the-code-cells) below and use the dropdown box or text box that appears to select a protein.\n",
    "\n",
    "To execute a cell, click on it and press ```Shift + Enter``` or the ```Run``` button in the notebook toolbar.\n",
    "\n",
    "Alternatively, run both cells by clicking ```Cell > Run All``` in the notebook toolbar.\n",
    "\n",
    "After selecting a protein, the protein and an accompanying menu will appear. The protein can be rotated by clicking and dragging. Scroll to zoom in and out. Different representations can be selected by navigating to the ```Extra > Quick``` tab in the protein viewer menu and selecting from the list of representations. A short reference for the protein menu can be found [here](https://github.com/zachmichael14/protein_viewer#gui-reference).\n",
    "\n",
    "# The Proteins <a class=\"anchor\" id=\"the-proteins\"></a>\n",
    "\n",
    "Proteins are all over and do all manner of important things, and that's an understatement. In fact, proteins are so important that the whole purpose of DNA is to serve as an instruction set for making them. \n",
    "\n",
    "Proteins are just chains of amino acids arranged in a certain way. The way these amino acid are arranged create the protein's structure, and this structure determines what the protein can do. \n",
    "\n",
    "For instance, a certain sequence of amino aicds might create a protein that has a small pocket for a chemical substrate to fit into, thus facilitating some reaction. An enzyme is just a protein that promotes some such reaction, though not necessarily in that way. Proteins with names ending in -ase are enzymes.\n",
    "\n",
    "You can enter a PDB ID below or select one of five proteins from the dropdown box. The five pre-selected proteins available here are:\n",
    "\n",
    "<ul>\n",
    "<li><strong>1C8Q: Human salivary amylase</strong><a class=\"anchor\" id=\"1c8q\"></a></li>\n",
    "    \n",
    "- Amylase is an enzyme that aids digestion by breaking down starches into sugars. It's made by both the pancreas and the salivary glands; its presence in the latter explains why eating foods high in starch (rice, for instance) may taste slightly sweet while being chewed.<br><br>\n",
    "    \n",
    "<li><strong>4M6U: *P. putida* mandelate racemase</strong><a class=\"anchor\" id=\"4m6u\"></a></li>\n",
    "    \n",
    "- The bacterium *Pseudomonas putida*, from which this protein comes, was the subject of the U.S. Supreme Court case *Diamond v. Chakrabarty* and ultimately became the first patented living organism. This particular protein essentially creates a mirror image of its substrate, a molecule called mandelic acid.<br><br>\n",
    "\n",
    "<li><strong>4ADS: Plasmodial PLP synthase complex</strong><a class=\"anchor\" id=\"4ads\"></a></li>\n",
    "    \n",
    "- The parasite *Plasmodium berghei* causes malaria in some rodents and is a model organism for studying malaria in humans. Protein 4ADS is a complex (meaning it's composed of multiple individual proteins) that synthesizes the molecule pyridoxal phosphate (PLP), or vitamin B<sub>6</sub>.<br><br>\n",
    "\n",
    "<li><strong>1UBQ: Ubiquitin (1.8 Å resolution)</strong><a class=\"anchor\" id=\"1ubq\"></a></li>\n",
    "    \n",
    "- Ubiquitin is so named because of its ubiquitous nature in most eukaryotes (animals, plants, and fungi). In fact, it's so useful that four diffent genes encode for its creation in humans. Importantly, ubiquitin fosters a process called ubiquintinylation, which can greatly alter a protein's function.<br><br>\n",
    "\n",
    "<li><strong>1ZQA: DNA polymerase complexed with DNA</strong><a class=\"anchor\" id=\"1zqa\"></a></li>\n",
    "    \n",
    "- DNA polymerases are a class of proteins that are cruicial in the replication and repair of DNA. This particular variant is involved heavily in the DNA repair process in humans. Here, it's pictured with a very small segment of DNA.\n",
    "</ul>\n",
    "\n",
    "\n",
    "# The code cells <a class=\"anchor\" id=\"the-code-cells\"></a>\n",
    "\n",
    "Run both of the cells below and select a protein from the dropdown menu or enter a PDB ID to view the protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f74f76c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" This code taken from ipywidget docs on debouncing.\n",
    "    The debounce is used for the protein textbox to avoid\n",
    "    attempting to retrieve incomplete PDB IDs.\n",
    "    https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20Events.html#debouncing \"\"\"\n",
    "import asyncio\n",
    "\n",
    "class Timer:\n",
    "    def __init__(self, timeout, callback):\n",
    "        self._timeout = timeout\n",
    "        self._callback = callback\n",
    "\n",
    "    async def _job(self):\n",
    "        await asyncio.sleep(self._timeout)\n",
    "        self._callback()\n",
    "\n",
    "    def start(self):\n",
    "        self._task = asyncio.ensure_future(self._job())\n",
    "\n",
    "    def cancel(self):\n",
    "        self._task.cancel()\n",
    "\n",
    "def debounce(wait):\n",
    "    \"\"\" Decorator that will postpone a function's\n",
    "        execution until after `wait` seconds\n",
    "        have elapsed since the last time it was invoked. \"\"\"\n",
    "    def decorator(fn):\n",
    "        timer = None\n",
    "        def debounced(*args, **kwargs):\n",
    "            nonlocal timer\n",
    "            def call_it():\n",
    "                fn(*args, **kwargs)\n",
    "            if timer is not None:\n",
    "                timer.cancel()\n",
    "            timer = Timer(wait, call_it)\n",
    "            timer.start()\n",
    "        return debounced\n",
    "    return decorator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f38a8d9-4482-48b4-bf7d-7a2fddde3ce9",
   "metadata": {
    "scrolled": false,
    "tags": [
     "remove-cell"
    ]
   },
   "outputs": [],
   "source": [
    "from os.path import exists\n",
    "from pathlib import Path\n",
    "\n",
    "from Bio.PDB import PDBList\n",
    "from Bio.PDB.MMCIFParser import MMCIFParser\n",
    "from Bio.PDB import MMCIF2Dict\n",
    "import ipywidgets as widgets\n",
    "import nglview as nv\n",
    "\n",
    "def visualize_protein(pdb_id, delete_file_after=False):\n",
    "    \"\"\" Parse structure from existing .cif file corresponding to pdb_id,\n",
    "        then view the protein using nglview GUI. Since get_pdb_file saves\n",
    "        the file locally, delete_file_after provides the option to\n",
    "        delete the saved file after viewing. \"\"\"\n",
    "    file = f'pdb_data/{pdb_id}.cif'\n",
    "    \n",
    "    # Parse molecular structure from file.\n",
    "    parser = MMCIFParser(QUIET=True)\n",
    "    structure = parser.get_structure(f'{pdb_id}', f'{file}')\n",
    "    \n",
    "    # Display the protein and nglview GUI.\n",
    "    view = nv.show_biopython(structure, gui=True)\n",
    "    with widget_output:\n",
    "        display(view)\n",
    "        # Clear output when new protein is selected.\n",
    "        widget_output.clear_output(wait=True)\n",
    "    \n",
    "    # Delete stored file if necessary.\n",
    "    if delete_file_after:\n",
    "        Path.unlink(file)\n",
    "\n",
    "def get_pdb_file(pdb_id, data_dir='pdb_data'):\n",
    "    \"\"\" Checks for .cif file in data_dir corresponding to pdb_id. \n",
    "        If it doesn't exist, retrieve it from PDB and save it locally. \n",
    "        Since the file is saved, this function returns nothing. \"\"\"\n",
    "    file_path = f'{data_dir}/{pdb_id}.cif'\n",
    "    \n",
    "    # If the file isn't stored, retrieve and save it locally.\n",
    "    if not exists(file_path):\n",
    "        pdb_list = PDBList()\n",
    "        \n",
    "        # TODO: Check for this to fail (no structure matching ID)\n",
    "        pdb_list.retrieve_pdb_file(pdb_id, pdir=\"pdb_data\")\n",
    "\n",
    "# SELECTOR WIDGETS\n",
    "pdb_dropdown = widgets.Dropdown(\n",
    "    options=[\n",
    "        'Select a protein...', \n",
    "        '1C8Q: Amylase ', \n",
    "        '4M6U: Racemase', \n",
    "        '4ADS: Synthase', \n",
    "        '1UBQ: Ubiquitin', \n",
    "        '1ZQA: Polymerase'\n",
    "    ],\n",
    "    value='Select a protein...',\n",
    "    description='PDB ID:',\n",
    "    disabled=False\n",
    ")\n",
    "\n",
    "pdb_textbox = widgets.Text(\n",
    "    value='',\n",
    "    placeholder='Enter a 4 character PDB ID...',\n",
    "    description='PDB ID:',\n",
    "    disabled=False\n",
    ")\n",
    "\n",
    "# SELECTOR EVENT HANDLERS\n",
    "def handle_dropdown(change):\n",
    "    # Handle error if user re-selects default selector value.\n",
    "    if change['new'] == 'Select a protein...':\n",
    "        with widget_output:\n",
    "            print('Please select a protein to display.')\n",
    "            widget_output.clear_output(wait=True)\n",
    "    else:\n",
    "        # Extract 4 character ID from selector widget.\n",
    "        pdb_id = change['new'][0:4].lower()\n",
    "        get_pdb_file(pdb_id)\n",
    "        visualize_protein(pdb_id)\n",
    "\n",
    "@debounce(1.5)\n",
    "def handle_textbox(change):\n",
    "    pdb_id = change['new'].lower()\n",
    "    \n",
    "    get_pdb_file(pdb_id)\n",
    "    visualize_protein(pdb_id)\n",
    "\n",
    "# Link selectors to event handlers.\n",
    "pdb_dropdown.observe(handle_dropdown, names='value')\n",
    "pdb_textbox.observe(handle_textbox, names='value')    \n",
    "    \n",
    "# Widget ouput must be captured in order to be displayed. \n",
    "widget_output = widgets.Output()\n",
    "\n",
    "# Display selector widget, output.\n",
    "display(widgets.VBox([pdb_dropdown, pdb_textbox, widget_output])) "
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
