{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a641a7f9",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    "<ol>\n",
    " <li><a href=\"#introduction\">Introduction</a></li>\n",
    " <ol>\n",
    "  <li><a href=\"#using-the-notebook\">Using the notebook</a></li>\n",
    " </ol>\n",
    "    <li><a href=\"#the-proteins\">The Proteins</a></li>   \n",
    "  <ol>\n",
    "    <li><a href=\"#1c8q\">1C8Q: Human salivary amylase</a></li>\n",
    "    <li><a href=\"#4m6u\">4M6U: *P. putida* mandelate racemase</a></li>\n",
    "    <li><a href=\"#4ads\">4ADS: Plasmodial PLP synthase complex</a></li>\n",
    "    <li><a href=\"#1ubq\">1UBQ: Ubiquitin (1.8 Å resolution)</a></li>\n",
    "    <li><a href=\"#1zqa\">1ZQA: DNA polymerase complexed with DNA</a></li>\n",
    "  </ol>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e426ec21",
   "metadata": {},
   "source": [
    "# Introduction <a class=\"anchor\" id=\"introduction\"></a>\n",
    "The Protein Data Bank (PDB) stores data for various proteins, each protein being designated by a four character alphanumeric ID. Some of this data includes 3-D structure information; this data is important because the structure of a protein determines its function.\n",
    "\n",
    "Even still, predicting the shape of a protein still remains challenging, though huge breakthroughs have been made in this field using [machine learning](https://pubmed.ncbi.nlm.nih.gov/34015749/). This notebook serves as a brief exploration into proteins and protein modeling.\n",
    "\n",
    "\n",
    "### Using the Notebook <a class=\"anchor\" id=\"using-the-notebook\"></a>\n",
    "To execute a cell, click on it and either press ```Shift + Enter``` or click the 'Run' button in the toolbar above. The cells are meant to be executed sequentially.\n",
    "\n",
    "After a cell is run, the ```In []:``` to the left of the cell should change to ```In [some number]:```.\n",
    "\n",
    "If you get any errors, check to make sure each cell has been executed. Otherwise, try reseting the kernel and running the cells again. To reset the kernel, click ```Kernel > Restart & Clear Output``` in the toolbar above.\n",
    "\n",
    "You can find more information about the notebook environment and how to troubleshoot issues [here](https://nbviewer.jupyter.org/github/ipython/ipython/blob/3.x/examples/Notebook/Index.ipynb).\n",
    "\n",
    "# The Proteins <a class=\"anchor\" id=\"the-proteins\"></a>\n",
    "\n",
    "Proteins are all over and do all manner of important things, and that's an understatement! They're so important in fact, that the whole purpose of DNA is to serve as an instruction set for making proteins. \n",
    "\n",
    "Proteins are just chains of amino acids arranged in a certain way. The way these amino acid are arranged create the protein's structure, and this structure determines what the protein can do. \n",
    "\n",
    "For instance, a certain sequence of amino aicds might create a protein that has small pocket for a chemical substrate to fit perfectly into, thus facilitating some reaction. In fact, an enzyme is just a protein that does exactly that, though not necessarily in that way. (note: if a protein's name ends in -ase, it's an enzyme).\n",
    "\n",
    "The available proteins are:\n",
    "\n",
    "<ul>\n",
    "<li><strong>1C8Q: Human salivary amylase</strong><a class=\"anchor\" id=\"1c8q\"></a></li>\n",
    "    \n",
    "- Amylase is an enzyme that aids digestion by breaking down starches into sugars. It's made by both the pancreas and the salivary glands; its presence in the latter explains why eating foods high in starch (rice, for instance) may taste slightly sweet while being chewed.\n",
    "    \n",
    "<li><strong>4M6U: *P. putida* mandelate racemase</strong><a class=\"anchor\" id=\"4m6u\"></a></li>\n",
    "    \n",
    "- The bacterium *Pseudomonas putida*, from which this protein comes, was the subject of the U.S. Supreme Court case *Diamond v. Chakrabarty* and ultimately became the first patented living organism . This particular protein essentially creates a mirror image of its substrate, a molecule called mandelic acid.\n",
    "\n",
    "<li><strong>4ADS: Plasmodial PLP synthase complex</strong><a class=\"anchor\" id=\"4ads\"></a></li>\n",
    "    \n",
    "- The parasite *Plasmodium berghei* causes malaria in some rodents and is a model organism for studying malaria in humans. Protein 4ADS is a complex (meaning it's composed of multiple individual proteins) that synthesizes the molecule pyridoxal phosphate (PLP), or vitamin B<sub>6</sub>.\n",
    "\n",
    "<li><strong>1UBQ: Ubiquitin (1.8 Å resolution)</strong><a class=\"anchor\" id=\"1ubq\"></a></li>\n",
    "    \n",
    "- Ubiquitin is so named because of its ubiquitous nature in most eukaryotes (animals, plants, and fungi). In fact, it's so useful that four diffent genes encode for its creation in humans. Importantly, ubiquitin fosters a process called ubiquintinylation, which can greatly alter a protein's function.\n",
    "\n",
    "<li><strong>1ZQA: DNA polymerase complexed with DNA</strong><a class=\"anchor\" id=\"1zqa\"></a></li>\n",
    "    \n",
    "- DNA polymerases are a class of proteins that are cruicial in the replication and repair of DNA. This particular variant is involved heavily in the DNA repair process in humans. Here, it's pictured with a very small segment of DNA.\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f38a8d9-4482-48b4-bf7d-7a2fddde3ce9",
   "metadata": {
    "scrolled": false,
    "tags": [
     "remove-cell"
    ]
   },
   "outputs": [],
   "source": [
    "# To display the selection widget,\n",
    "# press Shift + Enter or click the Run button in the toolbar above \n",
    "\n",
    "import warnings\n",
    "from Bio import BiopythonWarning\n",
    "from Bio.PDB import PDBParser\n",
    "import ipywidgets as widgets\n",
    "import nglview as nv\n",
    "\n",
    "# ignore Bio.PDBConstructionWarning for discontinuous protein chains\n",
    "warnings.simplefilter('ignore', BiopythonWarning)\n",
    "\n",
    "widget_output = widgets.Output() # capture widget output for display\n",
    "protein_selector = widgets.Dropdown(\n",
    "    options=[\n",
    "        'Select a protein...', \n",
    "        '1C8Q: Amylase ', \n",
    "        '4M6U: Racemase', \n",
    "        '4ADS: Synthase', \n",
    "        '1UBQ: Ubiquitin', \n",
    "        '1ZQA: Polymerase'\n",
    "    ],\n",
    "    value='Select a protein...',\n",
    "    description='PDB ID:',\n",
    "    disabled=False\n",
    ")\n",
    "\n",
    "def visualize_protein(change):\n",
    "    # handle error if user re-selects default selector value\n",
    "    if change['new'] == 'Select a protein...':\n",
    "        with widget_output:\n",
    "            print('Please select a protein to display.')\n",
    "            widget_output.clear_output(wait=True)\n",
    "    else:\n",
    "        pdb_id = change['new'][0:4] # get ID from selector widget\n",
    "        parser = PDBParser()\n",
    "        structure = parser.get_structure(f'{pdb_id}',f'pdb_data/{pdb_id}.pdb')\n",
    "        view = nv.show_biopython(structure, gui=True)\n",
    "        \n",
    "        with widget_output:\n",
    "            display(view)\n",
    "            widget_output.clear_output(wait=True)\n",
    "\n",
    "protein_selector.observe(visualize_protein, names='value')\n",
    "\n",
    "display(widgets.VBox([protein_selector, widget_output])) # display selector widget, output"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
